File: apps/api/app/schemas/notifications.py
from __future__ import annotations

from datetime import datetime
from typing import Literal
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError

from pydantic import BaseModel, Field, field_validator


NotificationFrequency = Literal["off", "daily", "weekdays", "weekly"]
NotificationSeverity = Literal["info", "attention", "important"]
NotificationKind = Literal[
    "alert",
    "watchlist",
    "calendar",
    "digest",
    "system",
]


class NotificationPreferences(BaseModel):
    in_app_enabled: bool = True
    email_enabled: bool = False
    digest_frequency: NotificationFrequency = "off"
    digest_time: str = "07:30"
    timezone: str = "America/Toronto"
    weekly_day: int = Field(default=0, ge=0, le=6)
    include_watchlist: bool = True
    include_portfolio: bool = True
    include_alerts: bool = True
    include_calendar: bool = True
    updated_at: datetime | None = None

    @field_validator("digest_time")
    @classmethod
    def validate_digest_time(cls, value: str) -> str:
        clean = value.strip()
        parts = clean.split(":")
        if len(parts) != 2:
            raise ValueError("L’heure doit utiliser le format HH:MM.")
        try:
            hour, minute = (int(part) for part in parts)
        except ValueError as error:
            raise ValueError("L’heure doit utiliser le format HH:MM.") from error
        if hour not in range(24) or minute not in range(60):
            raise ValueError("Heure de résumé invalide.")
        return f"{hour:02d}:{minute:02d}"

    @field_validator("timezone")
    @classmethod
    def validate_timezone(cls, value: str) -> str:
        clean = value.strip()
        try:
            ZoneInfo(clean)
        except ZoneInfoNotFoundError as error:
            raise ValueError("Fuseau horaire invalide.") from error
        return clean


class NotificationPreferencesEnvelope(BaseModel):
    preferences: NotificationPreferences
    account_email: str
    email_delivery_available: bool


class NotificationItem(BaseModel):
    id: str
    kind: NotificationKind
    title: str
    message: str
    severity: NotificationSeverity = "info"
    symbol: str | None = None
    route: str | None = None
    created_at: datetime
    read_at: datetime | None = None


class NotificationFeed(BaseModel):
    items: list[NotificationItem]
    unread_count: int = Field(ge=0)
    generated_at: datetime


class DigestSection(BaseModel):
    key: str
    title: str
    items: list[str] = Field(default_factory=list)


class NotificationDigest(BaseModel):
    subject: str
    greeting: str
    summary: str
    sections: list[DigestSection]
    generated_at: datetime
    disclaimer: str = (
        "Information générale seulement. Anatole ne formule aucune "
        "recommandation de placement."
    )


class DigestRunResult(BaseModel):
    processed: int = 0
    sent: int = 0
    skipped: int = 0
    failed: int = 0

File: apps/web/lib/notifications.ts
import { resilientFetch } from "./resilient-fetch";

export type NotificationFrequency = "off" | "daily" | "weekdays" | "weekly";

export type NotificationPreferences = {
  in_app_enabled: boolean;
  email_enabled: boolean;
  digest_frequency: NotificationFrequency;
  digest_time: string;
  timezone: string;
  weekly_day: number;
  include_watchlist: boolean;
  include_portfolio: boolean;
  include_alerts: boolean;
  include_calendar: boolean;
  updated_at: string | null;
};

export type NotificationPreferencesEnvelope = {
  preferences: NotificationPreferences;
  account_email: string;
  email_delivery_available: boolean;
};

export type NotificationItem = {
  id: string;
  kind: "alert" | "watchlist" | "calendar" | "digest" | "system";
  title: string;
  message: string;
  severity: "info" | "attention" | "important";
  symbol: string | null;
  route: string | null;
  created_at: string;
  read_at: string | null;
};

export type NotificationFeed = {
  items: NotificationItem[];
  unread_count: number;
  generated_at: string;
};

export type NotificationDigest = {
  subject: string;
  greeting: string;
  summary: string;
  sections: Array<{ key: string; title: string; items: string[] }>;
  generated_at: string;
  disclaimer: string;
};

async function notificationError(response: Response): Promise<Error> {
  try {
    const body = await response.json() as { detail?: string };
    if (body.detail) return new Error(body.detail);
  } catch {
    // Le statut HTTP reste visible si la réponse n'est pas JSON.
  }
  return new Error(`Erreur de notifications ${response.status}`);
}

async function notificationRequest<T>(
  path: string,
  options: RequestInit = {},
): Promise<T> {
  const headers = new Headers(options.headers);
  headers.set("Accept", "application/json");
  if (options.body) headers.set("Content-Type", "application/json");

  const response = await resilientFetch(`/api/notifications${path}`, {
    ...options,
    headers,
    credentials: "same-origin",
    retries: options.method && options.method !== "GET" ? 0 : 1,
    timeoutMs: 30_000,
    allowStale: false,
  });
  if (!response.ok) throw await notificationError(response);
  if (response.status === 204) return undefined as T;
  return await response.json() as T;
}

export function getNotificationFeed(): Promise<NotificationFeed> {
  return notificationRequest<NotificationFeed>("/feed?limit=80");
}

export function refreshNotifications(): Promise<NotificationFeed> {
  return notificationRequest<NotificationFeed>("/refresh", { method: "POST" });
}

export function markNotificationRead(id: string): Promise<void> {
  return notificationRequest<void>(`/feed/${encodeURIComponent(id)}/read`, {
    method: "POST",
  });
}

export function markAllNotificationsRead(): Promise<void> {
  return notificationRequest<void>("/read-all", { method: "POST" });
}

export function getNotificationPreferences(): Promise<NotificationPreferencesEnvelope> {
  return notificationRequest<NotificationPreferencesEnvelope>("/preferences");
}

export function saveNotificationPreferences(
  preferences: NotificationPreferences,
): Promise<NotificationPreferencesEnvelope> {
  return notificationRequest<NotificationPreferencesEnvelope>("/preferences", {
    method: "PUT",
    body: JSON.stringify(preferences),
  });
}

export function previewNotificationDigest(): Promise<NotificationDigest> {
  return notificationRequest<NotificationDigest>("/preview");
}

export function sendTestNotificationDigest(): Promise<NotificationDigest> {
  return notificationRequest<NotificationDigest>("/send-test", { method: "POST" });
}
